codalab/model/bundle_model.py

```diff
@@ -677,6 +677,58 @@ def get_bundle_uuids(self, conditions, max_results):
 
         return self._execute_query(query)
 
+    def get_memoized_bundles(self, user_id, command, dependencies):
+        # Select bundles that has the given command from the bundle table
+        filter_on_command = (
+            select([cl_bundle.c.uuid])
+            .select_from(cl_bundle)
+            .where(and_(cl_bundle.c.command == command, cl_bundle.c.owner_id == user_id))
+            .alias("filter_on_command")
+        )
+        # Get a list of uuids that matches with the given dependencies from the cl_bundle table.
+        # This step is to standardize uuid for future comparison.
+        dep_clause = [
+            cl_bundle_dependency.c.parent_uuid.like('%' + uuid + '%') for uuid in dependencies
+        ]
+        build_dependency_query = (
+            select([cl_bundle.c.uuid]).select_from(cl_bundle).where(or_(*dep_clause))
+        )
+        dependency_full_list = self._execute_query(build_dependency_query)
+        # Ensure the concatenated bundles' uuid in ascending order
+        dependency_full_list.sort()
+        concat_dependencies = ','.join(dependency_full_list)
+        # Join between filter_on_command table and cl_bundle_dependency table
+        # to get bundles that match with both input command and dependencies
+        join = (
```

Comment:
Are we considering the key for each dependency as well?  `cl run a:foo b:bar cat a b` is different from `cl run a:bar b:foo cat a b`

Reply:
Yes. That's a good catch! Currently, we don't have logic to handle this case. I'll add it.

Reply:
To resolve this use case, the solution I have is as follows:
1. pass `[a:uuid_a, b:uuid_b]` as a dependency list to `get_memoized_bundles` function.
1. resolve partial uuid in `uuid_a` and `uuid_b` and build a dependency string in ascending order.
1. query database between the `bundle` table and the `bundle_dependency` table based on **input command** and **the dependency list**, and build a dependency string in the same format as 2. Then select the uuids that have the same dependency string. 
1. Return the last uuid which indicates the most recent record from the result list. 
I think the current logic doesn't look really nice. I am still trying to come up with a better design in hoping to simplify this. But let me know if you have any suggestions.